### PR TITLE
Add syncidentityprovider controller.

### PR DESCRIPTION
**Pull request by twiest**
- [x] Add syncidentityprovider controller.
- [x] Add unit tests for syncidentityprovider controller

### CODE REVIEW

File: pkg/controller/syncidentityprovider/syncidentityprovider_controller.go [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+		// Wasn't a SelectorSyncIdentityProvider, bail out. This should not happen.
+		r.logger.Errorf("Error converting MapObject.Object to SelectorSyncIdentityProvider. Value: %+v", a.Object)
+		return retval
+	}
+
+	clusterDeployments := &hivev1.ClusterDeploymentList{}
+	r.List(context.TODO(), &client.ListOptions{}, clusterDeployments)
+
+	for _, clusterDeployment := range clusterDeployments.Items {
+		labelSelector, err := metav1.LabelSelectorAsSelector(&ssidp.Spec.ClusterDeploymentSelector)
```
Comment (csrwng) [inline pkg/controller/syncidentityprovider/syncidentityprovider_controller.go]: you should get this labelSelector once outside of this loop

File: pkg/controller/syncidentityprovider/syncidentityprovider_controller.go [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+
+	idps := []openshiftapiv1.IdentityProvider{}
+	for _, ssidp := range list.Items {
+		labelSelector, err := metav1.LabelSelectorAsSelector(&ssidp.Spec.ClusterDeploymentSelector)
+		if err != nil {
+			r.logger.WithError(err)
+			continue
+		}
+
+		if labelSelector.Matches(labels.Set(cd.Labels)) {
```
Comment (csrwng) [inline pkg/controller/syncidentityprovider/syncidentityprovider_controller.go]: minor optimization: create the labels.Set outside of this loop

File: pkg/controller/syncidentityprovider/syncidentityprovider_controller.go [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+	err := r.Client.List(context.TODO(), &client.ListOptions{Namespace: cd.Namespace}, list)
+	if err != nil {
+		return nil, err
+	}
+
+	idps := []openshiftapiv1.IdentityProvider{}
+	for _, sip := range list.Items {
+		for _, cdr := range sip.Spec.ClusterDeploymentRefs {
+			if cdr.Name == cd.Name {
+				idps = append(idps, sip.Spec.IdentityProviders...)
```
Comment (csrwng) [inline pkg/controller/syncidentityprovider/syncidentityprovider_controller.go]: break

File: pkg/controller/syncidentityprovider/syncidentityprovider_controller.go [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+	err := r.Client.List(context.TODO(), &client.ListOptions{}, list)
+	if err != nil {
+		return nil, err
+	}
+
+	idps := []openshiftapiv1.IdentityProvider{}
+	for _, ssidp := range list.Items {
+		labelSelector, err := metav1.LabelSelectorAsSelector(&ssidp.Spec.ClusterDeploymentSelector)
+		if err != nil {
+			r.logger.WithError(err)
```
Comment (csrwng) [inline pkg/controller/syncidentityprovider/syncidentityprovider_controller.go]: Missing log message

File: pkg/controller/syncidentityprovider/syncidentityprovider_controller.go [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      ssName,
+				Namespace: cd.Namespace,
+			},
+			Spec: *newSyncSetSpec,
+		}
+
+		// ensure the syncset gets cleaned up when the clusterdeployment is deleted
+		if err := controllerutil.SetControllerReference(cd, ss, r.scheme); err != nil {
+			r.logger.WithError(err).Error("error setting controller reference on job")
```
Comment (csrwng) [inline pkg/controller/syncidentityprovider/syncidentityprovider_controller.go]: s/job/syncset/

File: pkg/controller/syncidentityprovider/syncidentityprovider_controller.go [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+			return err
+		}
+
+		if err := r.Create(context.TODO(), ss); err != nil {
+			r.logger.Errorf("error creating sync set: %v", err)
```
Comment (csrwng) [inline pkg/controller/syncidentityprovider/syncidentityprovider_controller.go]: WithError

File: pkg/controller/syncidentityprovider/syncidentityprovider_controller.go [hunk 7]
```diff
@@ -0,0 +1,339 @@
...
+			return err
+		}
+
+		// we successfully created it.
+		return nil
+	}
+
+	if err != nil {
+		r.logger.Errorf("error checking for existing syncset: %v", err)
```
Comment (csrwng) [inline pkg/controller/syncidentityprovider/syncidentityprovider_controller.go]: WithError

File: pkg/controller/syncidentityprovider/syncidentityprovider_controller.go [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+	idpsFromSIDP, err := r.getRelatedSyncIdentityProviders(cd)
+	if err != nil {
+		return err
+	}
+
+	allIdps := append([]openshiftapiv1.IdentityProvider{}, idpsFromSSIDP...)
+	allIdps = append(allIdps, idpsFromSIDP...)
+
+	// Create a SyncSetSpec that includes all IdentityProviders as a patch
+	newSyncSetSpec, err := r.newSyncSetSpec(cd, allIdps)
```
Comment (csrwng) [inline pkg/controller/syncidentityprovider/syncidentityprovider_controller.go]: probably not a good idea to name the variable the same as the func (for maintainability)

File: pkg/controller/syncidentityprovider/syncidentityprovider_controller.go (hunk 3 above)
Comment (twiest) [reply pkg/controller/syncidentityprovider/syncidentityprovider_controller.go]: added

**Comment by twiest** [comment]
@csrwng This now ready to go. Please review when you get a minute. Thanks!

**Pull request merged**